### DISCUSSION TITLE
A Context API error in Next.js 14 app.

### DISCUSSION THREAD

**Discussion by the-pro7** [vercel/next.js#68249] [Help] [answered]
### Summary

I am getting this error `Error: (0 , _contexts_FormProvider__WEBPACK_IMPORTED_MODULE_5__.useFormData) is not a function`
when trying to use context values from a context I created like so: `const { typeOfForm, changeTypeOfForm } = useFormData()`
![Screenshot (1)](https://github.com/user-attachments/assets/1282c2c6-379b-4dff-a0a3-86d55b4b2949)

### Additional information

```js
Here's the the FormProvider.tsx code 
"use client";
import React, { createContext, useState, useContext } from "react";

type ContextType = {
  typeOfForm: "brandForm" | "influencerForm";
  changeTypeOfForm: () => void;
};

const FormContext = createContext<ContextType>({
  typeOfForm: "brandForm",
  changeTypeOfForm: () => {},
});

export const useFormData = () => useContext(FormContext);

const FormProvider: React.FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const [typeOfForm, setTypeOfForm] = useState<"brandForm" | "influencerForm">(
    "brandForm"
  );

  // Function to change type of form
  function changeTypeOfForm() {
    setTypeOfForm((prev) => {
      return prev == "brandForm" ? "influencerForm" : "brandForm";
    });
  }

  return (
    <FormContext.Provider value={{ typeOfForm, changeTypeOfForm }}>
      {children}
    </FormContext.Provider>
  );
};

export default FormProvider;


And here's how I'm using the `useFormData` in server component, `page.tsx`: ``
`const Page = () => {
  // Extract context values from Form Provider
  const { typeOfForm, changeTypeOfForm } = useFormData();
...
```
```


### Example

_No response_

**Comment by Joe-zzj** [comment]
Hi
The error you're encountering, "Error: (0 , contexts_FormProvider__WEBPACK_IMPORTED_MODULE_5_.useFormData) is not a function," typically indicates that useFormData is either not correctly imported or not properly used within a component that is wrapped by the FormProvider
Make sure that the FormProvider component is wrapping the component where useFormData is being called. For example, in a Next.js application, you can wrap the FormProvider around the main component in _app.tsx:
`// _app.tsx
import React from "react";
import FormProvider from "../contexts/FormProvider"; // Ensure the import path is correct

const MyApp: React.FC<{ Component: React.FC; pageProps: any }> = ({ Component, pageProps }) => {
  return (
    <FormProvider>
      <Component {...pageProps} />
    </FormProvider>
  );
};

export default MyApp;
`
Ensure that useFormData is correctly imported in your page.tsx file
Hope this helps!

**Reply by the-pro7** [reply]
Thanks, man, I guess not wrapping with the `FormProvider.tsx` was also a cause, but mainly because I did not have the `use client` directive.

**Comment by icyJoseph** [comment, marked as answer]
You page `[add-verification-info]/page.tsx`, also needs to have the `use client` directive at the top. 

It is false that not having a context provider wrapper would cause this error. When using context, if there's no provider for that context around the tree, then the default value is used. 

Since you have:

```js
const FormContext = createContext<ContextType>({
  typeOfForm: "brandForm",
  changeTypeOfForm: () => {},
});
```

That means, that using this context, without a FormProvider, gives you:

```js
{
  typeOfForm: "brandForm",
  changeTypeOfForm: () => {},
}
```

Of course be mindful of references.

Also the example given in the other comment, doesn't even utilize the FormProvider on the code snippet, and the code snippet is for _app, that is, Pages Router, whereas your screenshot shows you have the App Router.

Make sure the a layout is wrapping your `[add-verification-info]/page.tsx` page, is wrapping its children with your `FormProvider`.

**Reply by the-pro7** [reply]
Thanks pal it worked just fine!
